app/javascript/components/layer.jsx

```diff
@@ -24,187 +32,358 @@ export function Layer({ scope, layer, isOwner, layerId, activeTool, activeColor
+      const translatedDiffs = [];
+      data.indices.forEach((ind) => translatedDiffs.push({
+        'seq': tangibleSeqsRef.current[ind],
+        'data': layer.children[ind].exportJSON()
+      }));
+      layerChannel.send({
+        'diff_type': diffType,
+        'seq': seqRef.current++,
+        'data': {'translated_diffs': translatedDiffs, 'delta_x': data.delta_x, 'delta_y': data.delta_y}
+      });
     }
   }
 
   useEffect(() => {
     // Set up action cable subscriber once layer is created
     if (!!layer) {
-      setLayerChannel(setupSubscription(layer));
+      setLayerChannel(setupSubscription());
     }
   }, [layer]);
 
   useEffect(() => {
-    paperHandler();
+    // Once layer is channel is set up, call paperHandler (only do this for owner layer if owner or participant
+    // layer if participant)
+    if (!!scope) {
+      paperHandler();
+    }
   }, [layerChannel]);
 
   useEffect(() => {
+    // Transmit text diff if clicking out of a focused text box to change tool or color
+    if (!!pathRef.current && pathRef.current instanceof Paper.PointText) {
+      transmitDiff(DiffType.Tangible, pathRef.current);
+      pathRef.current.fullySelected = false;
+    }
     paperHandler();
-  }, [scope]);
+  }, [activeTool, activeColor]);
 
-  useEffect(() => {
-    if (pathRef.current) {
-      setPathState(oldPaths => [...oldPaths, pathRef.current]);
+  const eraseIntersectedItems = (path) => {
+    // Handles item removal via eraser
+    const indicesToRemove = [];
+    const removedSeqs = [];
+    layer.children.forEach((item, i) => {
+      // Get index and sequence number of removed items (layer.children has same order as tangibleSeqsRef)
+      if (path.id !== item.id && path.intersects(item)) {
+        removedSeqs.push(tangibleSeqsRef.current[i]);
+        indicesToRemove.push(i);
+      }
+    });
+    // Remove erased items
+    path.remove();
+    indicesToRemove.sort((a, b) => b - a);
+    indicesToRemove.forEach((i) => {
+      layer.children[i].remove();
+      tangibleSeqsRef.current.splice(i, 1);
+    });
+    // Create remove diff if any items were erased
+    if (removedSeqs.length > 0) {
+      transmitDiff(DiffType.Remove, removedSeqs);
     }
-  }, [activeTool, activeColor]);
+  };
 
-  useEffect(() => {
-    pathRef.current = null;
-    paperHandler();
-  }, [pathState.length, activeTool, activeColor]);
+  const deleteSelectedItems = () => {
+    // Handles item removal via select and delete
+    const indicesToRemove = [];
+    const removedSeqs = [];
+    scope.project.selectedItems.forEach((selectedItem) => {
+      // Get index and sequence number of removed items (layer.children has same order as tangibleSeqsRef)
+      let i = layer.children.findIndex((item) => selectedItem.id === item.id);
+      if (i >= 0) {
+        indicesToRemove.push(i);
+        removedSeqs.push(tangibleSeqsRef.current[i]);
+      }
+    });
+    // Remove selected items
+    indicesToRemove.sort((a, b) => b - a);
```

Comment:
question: do the indices need to be sorted?

Reply:
Yea so for this it essentially sorts `indicesToRemove` in descending order so that when we iterate over `indicesToRemove` it will remove the corresponding elements from the back to the front of the `layer.children[i]` and `tangibleSeqsRef.current` arrays (i.e., each removal won't affect indices earlier in the arrays). Could optimize by just keeping track of the number of elements already removed and doing `i - number of elements already removed` to get the index of the next element to delete but thought this was a bit simpler (not sure now that I think about it tho). What do you think?

Reply:
Actually ig can just iterate from back to front of list lol. Good call.